### ISSUE TITLE
Make setIn write to out-of-bounds array indices

### ISSUE THREAD

**Issue by ditkin** [HubSpot/transmute#34]
Running into bug where using `setIn` to write to out-of-bounds array indices just pushes to array instead.

Code: `setIn([2], 'abc', [])`
Expected: `[empty, empty, 'abc']`
Actual: `['abc']`

The cause is [the use of `splice()` in `_set`](https://github.com/HubSpot/transmute/blob/master/src/internal/_set.js#L9) when dealing with Arrays. Splice doesn’t let you write past the end of an array.

Solution:
instead of using `splice` we can use `next[index] = value`, which will overwrite existing values, or fill the array with `empty` to ensure we write to correct index.

**Comment by aem** [comment]
One option here to make this change in a backwards-compatible way would be to add this in the current version as a config option, so something like

```
setIn([2], 'abc', [], { fillMissingKeys: true });
```

which would result in your `[empty, empty, 'abc']` for explicit usages, while keeping the current behavior for existing users. Then if we ever need a new major we could flip that behavior, defaulting that flag to `true`

**Comment by kj800x** [comment]
I'm curious, how does this currently work in the case where we have an Immutable.js `List`? I believe when possible, it makes sense to have a consistent behavior between Lists and native arrays.

**Comment by ditkin** [comment]
@kj800x Good point - probably why `splice` was preferred over bracket notation for setting values.
